pkg/nvml/mock/shared/shared.go

```diff
@@ -0,0 +1,426 @@
+}
+
+// Server provides a reusable server implementation
+type Server struct {
+	mock.Interface
+	mock.ExtendedInterface
+	Devices           [8]nvml.Device
+	DriverVersion     string
+	NvmlVersion       string
+	CudaDriverVersion int
+}
+
+// Device provides a reusable device implementation
+type Device struct {
+	mock.Device
+	sync.RWMutex
+	UUID                  string
+	Name                  string
+	Brand                 nvml.BrandType
+	Architecture          nvml.DeviceArchitecture
+	PciBusID              string
+	Minor                 int
+	Index                 int
+	CudaComputeCapability CudaComputeCapability
+	MigMode               int
+	GpuInstances          map[*GpuInstance]struct{}
+	GpuInstanceCounter    uint32
+	MemoryInfo            nvml.Memory
+	PciDeviceId           uint32
+	MIGProfiles           MIGProfileConfig
+}
+
+// GpuInstance provides a reusable GPU instance implementation
+type GpuInstance struct {
+	mock.GpuInstance
+	sync.RWMutex
+	Info                   nvml.GpuInstanceInfo
+	ComputeInstances       map[*ComputeInstance]struct{}
+	ComputeInstanceCounter uint32
+	MIGProfiles            MIGProfileConfig
+}
+
+// ComputeInstance provides a reusable compute instance implementation
+type ComputeInstance struct {
+	mock.ComputeInstance
+	Info nvml.ComputeInstanceInfo
+}
+
+// CudaComputeCapability represents CUDA compute capability
+type CudaComputeCapability struct {
+	Major int
+	Minor int
+}
+
+var _ nvml.Interface = (*Server)(nil)
+var _ nvml.Device = (*Device)(nil)
+var _ nvml.GpuInstance = (*GpuInstance)(nil)
+var _ nvml.ComputeInstance = (*ComputeInstance)(nil)
+
+// NewServerFromConfig creates a new server from the provided configuration
+func NewServerFromConfig(config ServerConfig) *Server {
+	var devices [8]nvml.Device
+	for i := 0; i < config.GPUCount && i < 8; i++ {
+		devices[i] = NewDeviceFromConfig(config.Config, i)
+	}
+
+	server := &Server{
+		Devices:           devices,
+		DriverVersion:     config.DriverVersion,
+		NvmlVersion:       config.NvmlVersion,
+		CudaDriverVersion: config.CudaDriverVersion,
+	}
+	server.SetMockFuncs()
+	return server
+}
+
+// NewDeviceFromConfig creates a new device from the provided GPU configuration
+func NewDeviceFromConfig(config Config, index int) *Device {
+	device := &Device{
+		UUID:         "GPU-" + uuid.New().String(),
+		Name:         config.Name,
+		Brand:        config.Brand,
+		Architecture: config.Architecture,
+		PciBusID:     fmt.Sprintf("0000:%02x:00.0", index),
+		Minor:        index,
+		Index:        index,
+		CudaComputeCapability: CudaComputeCapability{
+			Major: config.CudaMajor,
+			Minor: config.CudaMinor,
+		},
+		GpuInstances:       make(map[*GpuInstance]struct{}),
+		GpuInstanceCounter: 0,
+		MemoryInfo:         nvml.Memory{Total: config.MemoryMB * 1024 * 1024, Free: 0, Used: 0},
+		PciDeviceId:        config.PciDeviceId,
+		MIGProfiles:        config.MIGProfiles,
+	}
+	device.SetMockFuncs()
```

Comment:
Note: Since a `device` is only functional after setting these mocks, should we be exposing it as a public concrete type?